packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/gridRowGroupingUtils.ts

```diff
@@ -76,34 +79,43 @@ export const filterRowTreeFromGroupingColumns = (
     areAncestorsExpanded: boolean,
   ): number => {
     let isMatchingFilters: boolean;
+    let isMatchingFilterItems: boolean;
+    let isMatchingQuickFilter: boolean;
     if (!isRowMatchingFilters) {
+      isMatchingFilterItems = true;
+      isMatchingQuickFilter = true;
       isMatchingFilters = true;
     } else {
       const shouldApplyItem = node.isAutoGenerated
         ? (columnField: string) => shouldApplyFilterItemOnGroup(columnField, node)
         : undefined;
 
-      isMatchingFilters = isRowMatchingFilters(node.id, shouldApplyItem);
+      const { passFilterItems, passQuickFilter } = isRowMatchingFilters(node.id, shouldApplyItem);
+      isMatchingFilterItems = passFilterItems;
+      isMatchingQuickFilter = passQuickFilter;
+      isMatchingFilters = isMatchingFilterItems && isMatchingQuickFilter;
     }
 
     let filteredDescendantCount = 0;
     node.children?.forEach((childId) => {
       const childNode = rowTree[childId];
       const childSubTreeSize = filterTreeNode(
         childNode,
-        areAncestorsPassingChildren && isMatchingFilters,
+        areAncestorsPassingChildren && isMatchingFilterItems,
         areAncestorsExpanded && !!node.childrenExpanded,
       );
       filteredDescendantCount += childSubTreeSize;
     });
 
-    let shouldPassFilters: boolean;
-    if (!areAncestorsPassingChildren) {
-      shouldPassFilters = false;
+    let shouldPassFilters = isMatchingFilters;
+    if (linkOperator === GridLinkOperator.And) {
```

Comment:
Not sure to understand what this code is doing

Reply:
Yes, it's over complicated. I modified it has follow:

- When filter logic is AND, I moved the early return a bit upper to avoid looking at the children
- Here I only distinguish the cases with/without tree structure
- renamed `shouldPassFilters` by `isPassingFilter`. The naming was a legacy of an experiment I did before reaching this solution